feat: Support for multi-aircraft type messages with selection of which one to show with reactions

pdellaert opened the pull request:
## feat: Support for multi-aircraft type messages with selection of which one to show with reactions

_****Check the video under Test Results for a comprehensive walkthrough of functionality****_

## Description

### Problem statements

* 104 out of 151 commands are all just simple embeds, no logic.
* Many of these commands contain A32NX specific information, but the command name itself could be very useful for the A380X in the future as well. 
* Goal is to avoid creating a whole set of new (simple message) commands for the A380X and re-use the same commands.

### Solution/Approach

Two major changes are introduced: 

* A new Command Definition interface is introduced: `MessageCommandDefinition`.
    * This is an interface for commands to implement that are just very simple messages and do not need a full executor (which is good for complex use cases)
    * It also comes with built-in support for the multi-aircraft capability by being able to configure messages for each type and a 'generic' message (which could introduce the choice in multi-aircraft commands, or just contain the command message for generic message commands)

* A new advanced command `typeCommand`.
    * This is a command with an executor, and is specifically designed to handle commands that have different messages per aircraft type. 
    * It registers itself with name/aliases that match the list of supported aircraft types (configured in the `constants`). 
    * When a aircraft type command is used with a sub-command that is actually a `MessageCommandDefinition`(example `.a32nx <regular command>`), the `messageCreate` handler executes the executor (as with any executor type command). The `typeCommand` executor will then look up the actual `MessageCommandDefinition` matching the sub-command, and post the embed matching the type used, if it exists (or not react if it doesn't).
    * When a simple `MessageCommandDefinition` command is executed by its regular command name (eg. `.airframe`), the `messageCreate` handler no longer executes the executor on the command itself (because it doesn't exist). Instead it executes the executor of the `typeCommand`, and because no aircraft type is given, it posts the generic message, which then shows a selection option with reactions of the supported aircraft types for that command (if it exists, if there isn't any, it just posts the generic message). 
    * It is possible to disable this type specific functionality, which is the case in the PR itself. This doesn't expose the `.<type>` commands and allows to keep the current behavior while porting all these simple message commands to `MessageCommandDefinitions`, where the `a32nx` message is being returned as the default. Once the type functionality is enabled, the generic message is used and the choices are offered if any available. 

This all allows for the following:

* Support multi-aircraft type commands and show specific data per aircraft type. 
* Do so by either directly calling the command for an aircraft type, or by offering the choice through reactions
* Simplify the basic message command management to no longer have to deal with an executor/posting. Normalizing and streamlining the creation of simple commands.
*  Provide controlled and gradual migration of existing commands, as well as adding information for other aircraft types without exposing that information in Discord. 
* Once the new aircraft type is ready, enable the functionality with the change of one line, simplifying the transition

Migration to `MessageCommandDefinition` from a `CommandDefinition` looks like this:

1. Change the interface to `MessageCommandDefinition`
2. Move the embed being posted in the executor to a variable outside of the `MessageCommandDefinition`. This variable can be called:
    1. `generic<command>Embed` in case this command doesn't need aircraft specific messages
    2. `a32nx<command>Embed` in case this command might need aircraft specific messages
3. Introduce a `generic<command>Embed` if this command needs aircraft specific messages
4. configure the `genericEmbed` property of the `MessageCommandDefinition` with the name of the `generic<command>Embed`
5. if it is a type specific command, set the `typeEmbed` property to `{ a32nx: a32nx<command>Embed}`. 
6. if it a type specific command, add a `a380x<command>Embed` with the a380x specific data, and also add that to the `typeEmbed`. 

Once the a380x has been released, the only thing to do is enable the functionality for aircraft specific messages in the `typeCommand`. 

## Test Results

Video, playback is at 1.3x because of time constraints with Youtube:
[![FBW Discord-bot Aircraft specific commands](https://img.youtube.com/vi/vrXiSlU2WdQ/0.jpg)](https://www.youtube.com/watch?v=vrXiSlU2WdQ)

## Discord Username

straks#7240

benw202 commented:
![image](https://user-images.githubusercontent.com/67422707/201524525-2ba1930b-0049-47c8-b2c2-b7dc692fa1e0.png)

Potential to add a column to command docs to denote which commands are 'split' and what 'types' the command has available.

benw202 commented:
A question about categories too. Obviously split commands are gonna be in 2 categories - do we build functionality for them to be in 2 categories?

They will show in the help embeds in those two categories, but I suppose that isn't a bad thing

benw202 commented:
MessageCommand vs ExecutorCommand.

MessageCommand is the new 'split' command type? Should we call it `splitCommandDefinition`? Is there a reason you have gone with message?

pdellaert commented:
> ![image](https://user-images.githubusercontent.com/67422707/201524525-2ba1930b-0049-47c8-b2c2-b7dc692fa1e0.png)
> 
> Potential to add a column to command docs to denote which commands are 'split' and what 'types' the command has available.

Great idea! But we could only update the doc once we actually have the a380x released, otherwise it might confuse people that the doc shows something that isn't available yet. So that would be a PR at the time of a380x release.


> A question about categories too. Obviously split commands are gonna be in 2 categories - do we build functionality for them to be in 2 categories?
> 
> They will show in the help embeds in those two categories, but I suppose that isn't a bad thing

I would rather suggest to rename the A32NX category to AIRCRAFT category, or something else generic instead. Adding multiple categories is not supported today, could be added, but seems like some work and extra logic in things like help command, for not that much benefit? 

We can expand the help command to include info on which aircraft type has specific info (once the feature is fully enabled). Let me think a little about that.

pdellaert commented:
Moved to draft because want to implement an improvement after discission with @benw202: 

Given the following scenario:

* starting point: both a32nx and a380x released and there's split commands that have embeds for each. 
* Now we plan to release a new aircraft (a3xxx) and we want to start creating embeds specific for that type. 
* Today's logic: the moment someone adds the a3xxx to the AircraftTypeList with an emoji, for any command we create a a3xxx embed, it will immediately show that option, because the split functionality would already be enabled (the general flag in the `typeCommand.ts`  file). 

Solution:
I can make that AircraftTypeList look like: 
```
export const AircraftTypeList = {
    a32nx: {
        reaction: '🛫',
        expose: true,
    },
    a380x: {
        reaction: '✈️',
        expose: true,
    },
    a3xxx: { // once we have it
        reaction: ...,
        expose: false,
    },
};
```
Even when the feature for multiple aircraft is enabled, it is still possible to hide one from the bot reaction list, even if an embed already exists for it. That way, we can add embeds for the new aircraft type in preparation for the release, but not yet expose them (as it would cause massive confusion).

benw202 commented:
> > ![image](https://user-images.githubusercontent.com/67422707/201524525-2ba1930b-0049-47c8-b2c2-b7dc692fa1e0.png)
> > Potential to add a column to command docs to denote which commands are 'split' and what 'types' the command has available.
> 
> Great idea! But we could only update the doc once we actually have the a380x released, otherwise it might confuse people that the doc shows something that isn't available yet. So that would be a PR at the time of a380x release.
> 

Agree on this - something to add in future though.

> > A question about categories too. Obviously split commands are gonna be in 2 categories - do we build functionality for them to be in 2 categories?
> > They will show in the help embeds in those two categories, but I suppose that isn't a bad thing
> 
> I would rather suggest to rename the A32NX category to AIRCRAFT category, or something else generic instead. Adding multiple categories is not supported today, could be added, but seems like some work and extra logic in things like help command, for not that much benefit?
> 
> We can expand the help command to include info on which aircraft type has specific info (once the feature is fully enabled). Let me think a little about that.

Yeah, aircraft will do! Then add info in help as you say

pdellaert commented:
> Moved to draft because want to implement an improvement after discission with @benw202:
> 
> Given the following scenario:
> 
> * starting point: both a32nx and a380x released and there's split commands that have embeds for each.
> * Now we plan to release a new aircraft (a3xxx) and we want to start creating embeds specific for that type.
> * Today's logic: the moment someone adds the a3xxx to the AircraftTypeList with an emoji, for any command we create a a3xxx embed, it will immediately show that option, because the split functionality would already be enabled (the general flag in the `typeCommand.ts`  file).
> 
> Solution: I can make that AircraftTypeList look like:
> 
> ```
> export const AircraftTypeList = {
>     a32nx: {
>         reaction: '🛫',
>         expose: true,
>     },
>     a380x: {
>         reaction: '✈️',
>         expose: true,
>     },
>     a3xxx: { // once we have it
>         reaction: ...,
>         expose: false,
>     },
> };
> ```
> 
> Even when the feature for multiple aircraft is enabled, it is still possible to hide one from the bot reaction list, even if an embed already exists for it. That way, we can add embeds for the new aircraft type in preparation for the release, but not yet expose them (as it would cause massive confusion).

I have a much easier solution:

* Only add the new option in `command.ts`, not in the `AircraftTypeList` in `constants.ts`
* Start adding the embeds to commands in prep for the new type
* Once released, update `AircraftTypeList`

This is demonstrated in the below video and only needed a little change in code (added in commit above)

[![FBW Discord-bot Aircraft specific commands Update 01](https://img.youtube.com/vi/6Pcc_WRppDE/0.jpg)](https://www.youtube.com/watch?v=6Pcc_WRppDE)

pdellaert commented:
Introduced:

* A way to add new aircraft without exposing it until it is released (see comment above)
* Renamed the category to a more generic type (Aircraft)

This is now ready for review

Note: Expanding help to show the supported types is planned for later in a separate PR.

benw202 commented on src/commands/support/discontinuity.ts:
```diff
@@ -5,7 +5,7 @@ import { makeEmbed, makeLines } from '../../lib/embed';
     name: ['discontinuity', 'disco'],
     description: 'Link to docs about discontinuities',
-    category: CommandCategory.A32NX,
+    category: CommandCategory.SUPPORT,
```
Any reason why this one moved to support rather than aircraft?

pdellaert replied:
It's in the support folder... So I fixed this one. We could move the entire command to the aircraft folder, though.

benw202 replied:
Ah right yes, I think that was probably an accident when command was created? I feel it should be in aircraft - what about you?

benw202 reviewed (commented):
A couple of comments, but this is looking really impressive! All functionality tested on my end and working well too. Nice work!

(typeCommand.ts - does it have to go in Utils?)

benw202 commented on src/commands/aircraft/airframe.ts:
```diff
@@ -0,0 +1,29 @@
+
+const genericAirframeEmbed = makeEmbed({
+    title: 'FlyByWire | simBrief Airframes',
+    description: 'Updated simBrief airframes exists for our different available aircraft. Please use the below reactions to select the correct airframe.',
```
```suggestion
    description: 'Updated simBrief airframes exist for our different available aircraft. Please use the below reactions to select the correct airframe.',
```

benw202 commented on src/commands/utils/typeCommand.ts:
```diff
@@ -0,0 +1,122 @@
+            const { data: genericEmbedData } = genericEmbed;
+            const postGenericEmbed = makeEmbed((genericEmbedData as EmbedData));
+            const choiceEmbedFieldLines = [];
+            choiceEmbedFieldLines.push('Please select the appropriate reaction for the Aircraft for which you would like more information:');
```
```suggestion
            choiceEmbedFieldLines.push('Please select the appropriate reaction for the Aircraft you would like more information about:');
```

Removes the double 'for'. Thoughts?

benw202 commented:
> > A couple of comments, but this is looking really impressive! All functionality tested on my end and working well too. Nice work!
> > We should ideally update `contributing.md` with this pr. Just to show how to create these new split commands - I don't think we need to explain in contributing how to create the 'regular' commands anymore as they will eventually be 'logic' only.
> > (typeCommand.ts - does it have to go in Utils?)
> 
> I'll work on `contributing.md`. We can move `typeCommand.ts` to another location. However, it does expose a set of actual commands (`.a32nx` and `.a380x` once the functionality is fully enabled)

Okay! Was thinking lib, but no biggie if you think not

pdellaert commented:
Latest changes:

- Using custom emojis from now on, this requires some special attention, so added comments in the constants.ts about this. Screenshot below of how it looks now
- Updated contributing doc example: https://github.com/pdellaert/discord-bot/blob/command_split_multiple_airframes/.github/CONTRIBUTING.md (also cleaned up some Markdown lint errors)
- Moved typeCommand to lib folder, as I agree it's a bit of a special command with complex logic. Because of this, I re-ordered the imports as well. 

<img width="605" alt="Screen Shot 2022-11-22 at 12 31 59 AM" src="https://user-images.githubusercontent.com/942391/203264363-30d2e85a-3e4f-4311-a879-1b9688925dac.png">

pdellaert commented:
Last update: 
Configured the _custom_ emojis present on the FBW Discord server (added today). 

**Warning**: 
Through this process I discovered the bot can only use custom emojis from the servers it is a member of. My test bot only runs on my personal server, and not on the FBW Discord, thus it doesn't have access to the custom emojis.
This fails the proper showing of the reactions and emojis by the bot (interestingly, even though I can use them manually on my server because I have access to both servers)

benw202 reviewed (approved):
LGTM
